src/Chainweb/Version.hs

```diff
@@ -209,109 +206,76 @@ instance IsMerkleLogEntry ChainwebHashTag ChainwebVersion where
+            TestWithTime{} -> "testWithTime"
+            TestWithPow{} -> "testWithPow"
+            Testnet00 -> "testnet00"
 {-# INLINABLE chainwebVersionToText #-}
 
--- | Read textual representation of Chainweb Version
+-- | Read textual representation of a `ChainwebVersion`.
 --
 chainwebVersionFromText :: MonadThrow m => T.Text -> m ChainwebVersion
-
--- Production versions
---
-chainwebVersionFromText "testnet00" = return Testnet00
-
--- Well-known test version names.
---
--- These are only used for parsing textual representations. There is a very low
--- chance that a roundtrip test for the 'HasTextRepresentation' of
--- 'ChainwebVersion' will succeed due to these names.
---
-chainwebVersionFromText "test" = return $ Test petersonChainGraph
-chainwebVersionFromText "test-singleton" = return $ Test singletonChainGraph
-chainwebVersionFromText "test-peterson" = return $ Test petersonChainGraph
-
-chainwebVersionFromText "testWithTime" = return $ TestWithTime petersonChainGraph
-chainwebVersionFromText "testWithTime-singleton" = return $ TestWithTime singletonChainGraph
-chainwebVersionFromText "testWithTime-peterson" = return $ TestWithTime petersonChainGraph
-
-chainwebVersionFromText "testWithPow" = return $ TestWithPow petersonChainGraph
-chainwebVersionFromText "testWithPow-singleton" = return $ TestWithPow singletonChainGraph
-chainwebVersionFromText "testWithPow-peterson" = return $ TestWithPow petersonChainGraph
-
--- Generic test versions
---
-chainwebVersionFromText t = case T.breakOnEnd "-" t of
-    (_, i) -> case treadM i of
-        Left e -> throwM
-            $ TextFormatException $ "Unknown Chainweb version: \"" <> t <> "\": " <> sshow e
-        Right x -> return $ fromTestChainwebVersionId x
-{-# INLINABLE chainwebVersionFromText #-}
+chainwebVersionFromText t =
+    case HM.lookup t chainwebVersions of
+        Just v -> pure v
+        Nothing -> case t of
+            "test" -> pure $ Test petersonChainGraph
+            "testWithTime" -> pure $ TestWithTime petersonChainGraph
+            "testWithPow" -> pure $ TestWithPow petersonChainGraph
+            _ -> throwM . TextFormatException $ "Unknown Chainweb version: " <> t
 
 instance HasTextRepresentation ChainwebVersion where
     toText = chainwebVersionToText
     {-# INLINE toText #-}
     fromText = chainwebVersionFromText
     {-# INLINE fromText #-}
 
+chainwebVersions :: HM.HashMap T.Text ChainwebVersion
+chainwebVersions = HM.fromList $
+    f Test "test"
+    <> f TestWithTime "testWithTime"
+    <> f TestWithPow "testWithPow"
+    <> [ ("testnet00", Testnet00) ]
+  where
+    f v p = map (\(k, g) -> (p <> k, v g)) pairs
+    pairs = [ ("-singleton", singletonChainGraph)
+            , ("-pair", pairChainGraph)
+            , ("-triangle", triangleChainGraph)
+            , ("-peterson", petersonChainGraph)
+            , ("-twenty", twentyChainGraph)
+            , ("-hoffman-singleton", hoffmanSingletonGraph)
+            ]
```

Comment:
It seems that this map is consulted each time a chainweb version is serialized or deserialized.

I am a bit concerned about mixing test and production versions here. In the previous version of the code only test versions uses special code paths. The lookup for production version was very straight forward and efficient.

What about keeping the old implementation of `chainwebVersionFromText` and just adapt the case

```
-- Generic test versions
	--
	chainwebVersionFromText t = ... 
```

for test versions?

Reply:
> It seems that this map is consulted each time a chainweb version is serialized or deserialized.

That old code also consulted the `TVar`'d Map for each serialization, too.

Reply:
Yes, that's my point: It did so *only* for test versions.